### ISSUE TITLE
Single Sign On with React app

### ISSUE THREAD

**Issue by janakapdj** [FusionAuth/fusionauth-issues#300]
Unable to sign out user and Verify 2nd application 

I have two separate applications with two login screens
in Application one  user login validated with fusionAurth with ( /api/login ) api rquest and it was success
but when I am trying to validate  same user for second Application (redirect with out log in)  with FusionAuth
I try with ( api/login/{userId}/{applicationId}?ipAddress={ipAddress} )
but this always return http status 200, even user log out 
and also user log out function  not work as expected 
( /api/logout?global={global}&refreshToken={refreshToken} )

help me with this 
Basically I would like to know Is it possible to control SSO with API and how its expire session on client application and how its authenticate user for second Application

**Comment by robotdan** [comment]
If you are using the Login API directly you will need to manage your own SSO. 

If you want to manage the tokens on your end to achieve SSO w/out using the FusionAuth SSO / OpenId Connect front end, you can do this using these APIs

- Login API https://fusionauth.io/docs/v1/tech/apis/login#authenticate-a-user
- JWT Issue API
-

**Comment by janakapdj** [comment]
Hi robotdan,
Thank for your reply ,
I am Not interest to manage SSO in My end, I am looking a way to manage SSO with Fusion Auth
just to clarify, below is way I plan to do it,
1. Login with API POST /api/login ( loginId , password)
2. If Login log in success call OAuth IdP login URL:   https://fusionauth.mydomain.com/oauth2/authorize?client_id .. to authorized user with Fusion auth
3. Log out all users  http://localhost:9011/oauth2/logout?..
Is this correct way to handle this ?
still I am not sure how to authorized user for second application with this

**Comment by robotdan** [comment]
The way FusionAuth manages SSO is with our login screens. So if you want SSO w/out our login screens, that means you are building it yourself. 

The APIs I referenced above do provide some of that function, such as exchange tokens between applications, etc. 

Why do you want to use the APIs directly s using the login screens, seems like the hard way. :-)

**Comment by janakapdj** [comment]
Hi robotdan,
Thank you again for your reply
I hope to use SAML v2 as a SSO solution is that most better one ?

**Comment by robotdan** [comment]
You can use SAML or OpenID Connect when taking advantage of our SSO, OpenID Connect is preferred if you are able to use it. 

To test with https://samltest.id you will need to first upload your metadata. Make sure you are running FusionAuth version 1.7.4 or newer. 

1. Go to https://samltest.id/upload.php
2. Enter your Entity Id URL in the first field and click on "Fetch!"

You will find your Entity Id URL in the `SAML v2 Integration details` section of the Application details. To view these details navigate to `Applications` from the left nav and click on the green view button. The Entity Id URL (along with the Metadata URL) will return an XML document that can be used by the SP (in this case https://samltest.id) to verify your IdP.

**Comment by janakapdj** [comment]
Hi robodan,

I have Installed latest version 

I was able to create XML file , but now stuck with Idp url ,

again ill clear my requirement 

I have one application done with react and other one is third party app that support SAML(but can not do any change with source )
All the users for both applications has imported in to Fusion auth (two fusion auth applications)  
users need to allow to access both applications with SSO
Is this possible to achieve ? or do I need to setting up Idp server ?
sorry for asking questions again and again :)

**Comment by robotdan** [comment]
> Is this possible to achieve ? or do I need to setting up Idp server ?

I believe so, as long as our SAML implementation is compatible with yours. 

> I was able to create XML file , but now stuck with Idp url ,

I don't know what that means, how can I help you?

**Comment by janakapdj** [comment]
Hi 
I push back to SSO with locally runing web app, and i am trying to connect fusion with sisense is there any documentation how to connect with sisense ? in sisense I found documentations for some other 3rd part apps but could not find for fusion :(

**Comment by janakapdj** [comment]
I am running on a docker here is a log file

```Sep 27, 2019 5:29:40.411 AM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
java.lang.IllegalArgumentException: Invalid SAML v2.0 Name ID format [urn:oasis:names:tc:SAML:2.0:nameid-format:emailAddress]
        at io.fusionauth.samlv2.domain.NameIDFormat.fromSAMLFormat(NameIDFormat.java:156)
        at io.fusionauth.samlv2.service.DefaultSAMLv2Service.parseRequest(DefaultSAMLv2Service.java:471)
        at io.fusionauth.app.action.samlv2.LoginAction.get(LoginAction.java:92)
        at sun.reflect.GeneratedMethodAccessor122.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.primeframework.mvc.util.ReflectionUtils.invoke(ReflectionUtils.java:436)
        at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.execute(DefaultActionInvocationWorkflow.java:84)
        at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.perform(DefaultActionInvocationWorkflow.java:64)
        at

**Comment by robotdan** [comment]
I don't believe that is a valid Name Id Format. The email address `NameID` format is `urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress`.  

What SAML provider are you using that is using that Name ID Format? 

>  What should be IDP endpoint ? its able to create with fusion or some other Api ?

If you click the view button you will see the SAML integration details you will see a Login URL and an EntityID URL.

**Comment by janakapdj** [comment]
Hi robotdan,
Thank you again for your reply
I am still getting same error on log file , as I seen NameID format is correct 
for your reference I have attached some screen captures , I am passing my dead line and need to get done with fusion or alternative :(
![SAML](https://user-images.githubusercontent.com/5866239/65849553-b1c39680-e368-11e9-8432-f759d2ace082.PNG)
![Application](https://user-images.githubusercontent.com/5866239/65849558-b5efb400-e368-11e9-91df-c0284aa99853.PNG)

**Comment by robotdan** [comment]
As far as I know that NameId format is not valid, however I am sure there are IdPs out there using it - case in point. 

For this reason, in the latest release of FusionAuth we will try to guess the NameId format even if it doesn't appear to be one that we recognize. In this particular case, the NameId Format of `urn:oasis:names:tc:SAML:2.0:nameid-format:emailAddress` that is being sent to FusionAuth will resolve to the NameId format for email. 

Let me know if the latest version of FusionAuth works for you.

**Comment by janakapdj** [comment]
Hi robotdan,
thank for your reply/suggestion
I have re installed latest version now its seems NameId format error has fixed ,
now its given a error 
**"message" : "The AuthnRequest contained an invalid issuer [Sisense] that does not map to an Application in FusionAuth"**
this not print any error on log file
 as I understood Issuer given for application is not correct , if it is how I find a correct issuer ?
I sissense document there is no reference related to this
![error](https://user-images.githubusercontent.com/5866239/66027903-892ece80-e519-11e9-9dc2-b812ba995385.PNG)

**Comment by janakapdj** [comment]
Hi robotdan,
when I try to login in to  Sisense  "http://localhost:8081" now its redirect to Fusion auth login screen
(login screen attached) 
but after fill credentials and click in login button Fusion auth its redirect again to Sisesne login screen
When click on Login With Sisense button it shows XML content  

and also I seen this error on developer tool when we enter credentials on Fusion auth login 
content_script.js:62 LEVEL: ERROR | MODULE: LEAKED_CREDENTIALS | SESSION: ac1cc6e9 | MESSAGE: "CryptoJS.SHA256 is not a function" | PARAMS: {}

![login](https://user-images.githubusercontent.com/5866239/66035963-c69b5800-e529-11e9-9014-e940a81675b4.PNG)

**Comment by robotdan** [comment]
Hi @janakapdj I would need to see your SAML configuration to help much with this. Also, enable debug in your SAML configuration and capture the Debug event log.

**Comment by janakapdj** [comment]
Hi @robotdan here I have attached screen shots Please let me know if you need more information on this
![app](https://user-images.githubusercontent.com/5866239/66180396-8b646a80-e68a-11e9-9956-2dc6614e7122.png)
![app_saml](https://user-images.githubusercontent.com/5866239/66180398-8dc6c480-e68a-11e9-9085-3666e8ed5c69.png)
![application](https://user-images.githubusercontent.com/5866239/66180406-928b7880-e68a-11e9-9eb0-60d1f368fe3f.PNG)
![saml](https://user-images.githubusercontent.com/5866239/66180409-961eff80-e68a-11e9-8a93-a7ebf18f4120.png)

**Comment by janakapdj** [comment]
@robotdan could you help me with this

**Comment by robotdan** [comment]
> and also I would like to know does fusion auth support to external https request


Are you asking if we support AJAX inside of a lambda function? Currently we do not support this. 
( see https://github.com/FusionAuth/fusionauth-issues/issues/267 )

> Also, enable debug in your SAML configuration and capture the Debug event log.

Can you provide any related Error or Debug event logs that are produced? 


> content_script.js:62 LEVEL: ERROR | MODULE: LEAKED_CREDENTIALS | SESSION: ac1cc6e9 | MESSAGE: "CryptoJS.SHA256 is not a function" | PARAMS: {}

This is not part of FusionAuth, is this `crypto-js` ? 

https://www.npmjs.com/package/crypto-js
https://stackoverflow.com/questions/51953592/sha256-is-undefined

**Comment by janakapdj** [comment]
Hi,
```
When click on Login With Sisense button it shows XML content 
Can you provide any related Error or Debug event logs that are produced?
```
here not generate any event log I have enabled debug on application and SAML configurations too
for your reference I have attached screen recorder 
I check in docker log file too that also not create any log entry
[Oct 11, 2019 9_54 AM.zip](https://github.com/FusionAuth/fusionauth-issues/files/3715648/Oct.11.2019.9_54.AM.zip)

**Comment by robotdan** [comment]
Here is the node client link in NPM, https://www.npmjs.com/package/@fusionauth/node-client 

This library provides access to our APIs, SSO is user agent based (web browser) so the only part that you'll need to use in this library is the `exchangeRefreshTokenForJWT` method.

**Comment by janakapdj** [comment]
** In your screenshot your clicking the Login with SAML button which I assume means you want to log into Sisense via SAML v2. : Yes 
here I attache SAML v2 settings 
I try  changing  callback url / issuer  in Application but ended with XML response :(
![saml](https://user-images.githubusercontent.com/5866239/66804658-aebbcf00-ef40-11e9-8d9a-012ef619fb73.png)
![application](https://user-images.githubusercontent.com/5866239/66804660-afecfc00-ef40-11e9-885c-c9d391183b89.png)
![idp](https://user-images.githubusercontent.com/5866239/66804662-b11e2900-ef40-11e9-8924-d3e0d5cb1831.png)

**Comment by robotdan** [comment]
I may have mispoke, it looks like the Node client doesn't have the method I was thinking of..  here is the endpoint you can call it via Node relatively easily. 
https://fusionauth.io/docs/v1/tech/oauth/endpoints#complete-the-authorization-code-grant-request

> i my case with node client authentication is success but still Sisense (configured with SAML) redirecting to login screen but it should be redirecting to sisense dashboard . is there any additional configurations to do ?
or its not possible to do SSO with node client ? because its make authentication through API ?

This is correct, as I've stated several times, when you use our API you are not using our SSO. You need to use our login pages for SSO to work how you'd like. 

Using our login pages instead of building it in your app is a better strategy for security, and effort on your end as well. You should be able to theme our login to look just like your site. 

For reference
https://fusionauth.io/articles/logins/types-of-logins-authentication-workflows

**Comment by janakapdj** [comment]
Hi robotdn,

i am working on authenticated node appliction with oAuth
when its try to get toke with authnticated code i get below 

{error: "invalid_client",…}
error: "invalid_client"
error_description: "Client authentication missing as Basic Authorization header or credentials in the body (or some combination of them)."
error_reason: "client_authentication_missing"

Axios post :
Request URL: http://10.197.55.100:9011/oauth2/token
Request Method: POST
Status Code: 401 
Remote Address: 10.197.55.100:9011
Referrer Policy: no-referrer-when-downgrade
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: http://localhost:3000
Access-Control-Expose-Headers: Access-Control-Allow-Origin,Access-Control-Allow-Credentials
Content-Length: 213
Content-Type: application/json;charset=UTF-8
Date: Wed, 16 Oct 2019 12:02:24 GMT
Vary: Origin
Provisional headers are shown
Accept: application/json, text/plain, */*
Content-Type: application/x-www-form-urlencoded
Origin: http://localhost:3000
Referer: http://localhost:3000/Authenticated?code=2BKjKK9195Vl-PBNujcRUAKA18Its3hGyvIasENrVgQ&locale=en_US&userState=Authenticated
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.120 Safari/537.36
client_id: c6bcfb81-7387-4448-92fe-979fbc183864
code: 2BKjKK9195Vl-PBNujcRUAKA18Its3hGyvIasENrVgQ
grant_type: authorization_code
redirect_uri: http://localhost:3000/Authenticated

**Comment by robotdan** [comment]
Have you reviewed the Token endpoint doc? 
https://fusionauth.io/docs/v1/tech/oauth/endpoints#complete-the-authorization-code-grant-request


You either need to provide and `Authorization` header or disable client authentication for this FusionAuth Application.

![image](https://user-images.githubusercontent.com/1148709/66945339-1e4ac080-f00c-11e9-97bd-c7b1b641066f.png)

**Comment by robotdan** [comment]
I think there are a bunch of issues we're working through on this single ticket so it is difficult to follow where we are at. 

I'll close out this issue, and open a new issue if you still need assistance with a specific problem.
